File: ts/instructions/setRoot.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface SetRootArgs {
  admin: PublicKey
  treeCreationFee: BN
}

export interface SetRootAccounts {
  signer: PublicKey
  /** The global root */
  root: PublicKey
  /** Common Solana programs */
  tokenProgram: PublicKey
  associatedTokenProgram: PublicKey
  systemProgram: PublicKey
  rent: PublicKey
}

export const layout = borsh.struct([
  borsh.publicKey("admin"),
  borsh.u64("treeCreationFee"),
])

export function setRoot(args: SetRootArgs, accounts: SetRootAccounts) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.root, isSigner: false, isWritable: false },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.rent, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([183, 49, 10, 206, 168, 183, 131, 67])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      admin: args.admin,
      treeCreationFee: args.treeCreationFee,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: programs/dippies_index_protocol/src/instructions/set_root.rs
use anchor_lang::prelude::*;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::Token;

use crate::seeds::ROOT_SEED;
use crate::state::Root;

pub fn set_root(ctx: Context<SetRoot>, admin: Pubkey, tree_creation_fee: u64) -> Result<()> {
    msg!("Setting the global root");

    let root = &mut ctx.accounts.root;

    root.admin = admin;
    root.tree_creation_fee = tree_creation_fee;

    Ok(())
}

#[derive(Accounts)]
#[instruction(admin: Pubkey)]
pub struct SetRoot<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// The global root
    #[account(
        seeds = [
            ROOT_SEED.as_bytes(),
            &root.id.to_bytes(),
        ],
        bump,
        has_one = admin,
    )]
    pub root: Account<'info, Root>,

    /// Common Solana programs
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}
